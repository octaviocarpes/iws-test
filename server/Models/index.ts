export { Album } from './Album';
export { Band } from './Band';
