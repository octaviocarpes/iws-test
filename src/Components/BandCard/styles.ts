import { createUseStyles } from 'react-jss';
import { colors } from '../../styles/colors';

export const useStyles = createUseStyles({
  card: {
    border: ['1px solid lightgray'],
    boxShadow: ['0px 3px 10px 0px rgb(224, 220, 224)'],
    borderRadius: 20,
    margin: 10,
    width: '100%',
    height: 100,
    display: 'flex',
    justifyContent: 'space-between',

    '& img': {
      height: 75,
      width: 75,
      borderRadius: 10,
    },

    '& p': {
      fontFamily: 'Montserrat',
    },
  },

  clicked: {
    border: ['1px solid lightgray'],
    borderRadius: 20,
    margin: 10,
    width: '100%',
    height: 100,
    display: 'flex',
    justifyContent: 'space-between',

    '& img': {
      height: 75,
      width: 75,
      borderRadius: 10,
    },

    '& p': {
      fontFamily: 'Montserrat',
    },
    boxShadow: ['none'],
  },

  cardWrapper: {
    width: '50%',
    display: 'flex',
    justifyContent: 'flex-start',
    alignItems: 'center',
  },

  bandImage: {
    display: 'flex',
    justifyContent: 'flex-start',
    alignItems: 'center',
    paddingLeft: 10,
    paddingRight: 10,
  },

  name: {
    fontWeight: 'bold',
    color: colors.primaryGray,
    whiteSpace: 'nowrap',
  },

  genre: {
    fontWeight: 'bold',
    fontSize: 11,
    textTransform: 'capitalize',
    color: colors.primaryOrange,
  },

  bandInfo: {
    display: 'flex',
    flexDirection: 'column',
    justifyContent: 'space-around',
    alignItems: 'left',
  },

  bandPlays: {
    display: 'flex',
    flexDirection: 'column',
    justifyContent: 'flex-end',
    paddingRight: 15,
  },

  plays: {
    fontSize: 12,
    fontWeight: 'bold',
    color: colors.primaryGray,
  },
});
