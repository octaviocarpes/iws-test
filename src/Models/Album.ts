export interface Album {
  name: string;
  image: string;
  releasedDate: Date;
  band: string;
  tracks: [
    {
      name: string;
      duration: number;
      id: string;
    },
  ];
  id: string;
}
