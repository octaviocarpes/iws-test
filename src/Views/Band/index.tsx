import React, { useState, useEffect, useContext } from 'react';
import { observer } from 'mobx-react';
import { Band } from '../../Models/Band';
import { getBandById, getBandAlbums } from '../../Services/BandsService';
import { LoadingContext } from '../../Contexts/Loading.context';

import Image from '../../Components/Image';

import fallbackImage from '../../styles/img/album-icon.png';
import { useStyles } from './styles';

const BandsView = observer(props => {
  const loadingStore = useContext(LoadingContext);
  const [band, setBand] = useState<Band>();
  const styles = useStyles();

  const getBandInfo = async (): Promise<Band> => {
    try {
      const bandResponse = await getBandById(props.id);
      const albumsResponse = await getBandAlbums(props.id);
      const band: Band = {
        ...bandResponse.data,
      };
      band.albums = albumsResponse.data;
      console.log(band);
      return band;
    } catch (error) {
      return error;
    }
  };

  useEffect(() => {
    if (band === undefined) {
      loadingStore.setIsLoading(true);
      getBandInfo()
        .then(response => {
          setBand(response);
          loadingStore.setIsLoading(false);
        })
        .catch(error => {
          // TODO: notify error,
          console.log(error);
        });
    }
  }, [band]);

  return (
    <>
      {band !== undefined ? (
        <div className={styles.bandView}>
          <Image src={band.image} fallbackSrc={fallbackImage} alt="Band Image" />
        </div>
      ) : (
        <></>
      )}
    </>
  );
});

export default BandsView;
