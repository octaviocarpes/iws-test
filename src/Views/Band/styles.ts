import { createUseStyles } from 'react-jss';
import { colors } from '../../styles/colors';

export const useStyles = createUseStyles({
  bandView: {
    display: 'flex',
    flexDirection: 'column',
    justifyContent: 'center',
    alignItems: 'center',
    width: '100%',
    height: '100%',
    fontFamily: 'Montserrat',
  },

  goBack: {
    position: 'absolute',
    top: 0,
    left: 0,
    width: 100,
    height: 40,
    marginTop: 15,
    marginLeft: 15,

    '& button': {
      width: '100%',
      height: '100%',
      background: 'white',
      fontSize: 16,
      fontFamily: 'Montserrat',
      fontWeight: 'bold',
      color: colors.primaryOrange,
      border: [`1px solid ${colors.primaryOrange}`],
      borderRadius: 20,

      '&:focus': {
        outline: 'none',
      },
    },
  },

  bandLogo: {
    border: ['1px solid black'],
    width: '100%',
    height: 400,

    '& img': {
      width: '100%',
      height: 400,
    },
  },

  bandHeader: {
    display: 'flex',
    flexDirection: 'column',
    justifyContent: 'center',
    alignItems: 'center',

    '& h1': {
      color: colors.primaryGray,
    },

    '& h3': {
      color: colors.primaryOrange,
      textTransform: 'capitalize',
    },
  },

  biography: {
    display: 'flex',
    flexDirection: 'column',
    justifyContent: 'center',
    alignItems: 'center',
    padding: 15,

    '& p': {
      color: colors.primaryGray,
      textAlign: 'justify',
      lineHeight: 1.8,
      fontSize: 15,
    },

    '& a': {
      color: colors.primaryOrange,
    },
  },

  albums: {},
});
