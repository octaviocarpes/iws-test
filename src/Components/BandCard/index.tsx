import React, { memo, ReactElement, useState } from 'react';
import { Band } from '../../Models/Band';
import { useStyles } from './styles';
import Image from '../Image';
import fallbackImage from '../../styles/img/album-icon.png';

interface Props {
  band: Band;
  onClick?: () => void;
}

const BandCard = memo(
  (props: Props): ReactElement => {
    const { band, onClick } = props;
    const styles = useStyles();
    const [cardStyle, setCardStyle] = useState(styles.card);

    const _handleTouchStart = (): void => {
      setCardStyle(styles.clicked);
    };

    const _handleTouchEnd = (): void => {
      setCardStyle(styles.card);
      if (onClick) onClick();
    };

    return (
      <>
        <div
          className={cardStyle}
          onTouchStart={(): void => _handleTouchStart()}
          onTouchEnd={(): void => _handleTouchEnd()}
        >
          <div className={styles.cardWrapper}>
            <div className={styles.bandImage}>
              <Image alt={`${band.name}-image`} src={band.image} fallbackSrc={fallbackImage} />
            </div>
            <div className={styles.bandInfo}>
              <p className={styles.name}>{band.name}</p>
              <p className={styles.genre}>{band.genre}</p>
            </div>
          </div>

          <div className={styles.bandPlays}>
            <p className={styles.plays}>Plays: {band.numPlays}</p>
          </div>
        </div>
      </>
    );
  },
);

BandCard.displayName = 'BandCard';
export default BandCard;
