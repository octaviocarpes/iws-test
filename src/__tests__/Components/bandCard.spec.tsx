import React from 'react';
import { render, cleanup } from '@testing-library/react';
import BandCard from '../../Components/BandCard';
import { Band } from '../../Models/Band';

const band: Band = {
  name: 'Band',
  id: '12345',
  image: 'image',
  albums: [],
  biography: 'Biography',
  genre: 'genre',
  numPlays: 10000,
};

afterEach(cleanup);

describe('<BandCard />', () => {
  it('should contains the bands name and genre', () => {
    const { getByText } = render(<BandCard band={band} />);
    getByText('Band');
    getByText('genre');
  });

  it('should render the component correctly', () => {
    const tree = render(<BandCard band={band} />);
    expect(tree).toMatchSnapshot();
  });
});
