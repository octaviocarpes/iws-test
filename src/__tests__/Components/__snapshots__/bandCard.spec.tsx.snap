// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<BandCard /> should render the component correctly 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="card-0-2-8"
      >
        <div
          class="cardWrapper-0-2-1"
        >
          <div
            class="bandImage-0-2-2"
          >
            <img
              alt="Band-image"
              src="image"
            />
          </div>
          <div
            class="bandInfo-0-2-5"
          >
            <p
              class="name-0-2-3"
            >
              Band
            </p>
            <p
              class="genre-0-2-4"
            >
              genre
            </p>
          </div>
        </div>
        <div
          class="bandPlays-0-2-6"
        >
          <p
            class="plays-0-2-7"
          >
            Plays: 
            10000
          </p>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="card-0-2-8"
    >
      <div
        class="cardWrapper-0-2-1"
      >
        <div
          class="bandImage-0-2-2"
        >
          <img
            alt="Band-image"
            src="image"
          />
        </div>
        <div
          class="bandInfo-0-2-5"
        >
          <p
            class="name-0-2-3"
          >
            Band
          </p>
          <p
            class="genre-0-2-4"
          >
            genre
          </p>
        </div>
      </div>
      <div
        class="bandPlays-0-2-6"
      >
        <p
          class="plays-0-2-7"
        >
          Plays: 
          10000
        </p>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
